tensorboard/data/server/reservoir.rs

```diff
@@ -234,6 +236,55 @@ impl<T, C: ReservoirControl> StageReservoir<T, C> {
         &self.staged_items[..]
     }
 
+    /// Checks whether a preemption has been staged.
+    ///
+    /// This is initially `false`; it changes to `true` whenever a record is preempted, and back to
+    /// `false` whenever the reservoir is committed.
+    pub fn staged_preemption(&self) -> bool {
+        self.staged_preemption
+    }
+
+    /// Preempts any records whose step does not precede the given step.
+    fn preempt(&mut self, step: Step) {
+        let staged_nonpreempted = self
+            .staged_items
+            .iter()
+            .enumerate()
+            .rev()
+            .find_map(|(i, (s, _))| if *s < step { Some(i + 1) } else { None })
+            .unwrap_or(0);
+        let staged_preempted = self.staged_items.len() - staged_nonpreempted;
+        if staged_preempted > 0 {
+            self.staged_items.truncate(staged_nonpreempted);
+            self.staged_preemption = true;
+        }
+        let mut preempted = staged_preempted;
+        if self.staged_items.is_empty() {
+            // Committed steps may have been preempted as well.
+            let committed_nonpreempted = self
+                .committed_steps
+                .iter()
+                .enumerate()
+                .rev()
+                .find_map(|(i, s)| if *s < step { Some(i + 1) } else { None })
+                .unwrap_or(0);
+            let committed_preempted = self.committed_steps.len() - committed_nonpreempted;
+            if committed_preempted > 0 {
+                self.committed_steps.truncate(committed_nonpreempted);
+                self.staged_preemption = true;
+                preempted += committed_preempted;
+            }
+        }
+        if preempted == 0 {
+            return; // No need to adjust `seen`.
+        }
+        let old_stored = self.committed_steps.len() + self.staged_items.len() + preempted;
```

Comment:
Seems a little more self-evident to just do `let old_stored = self.len()` at the top of the function?  Then I think we might also not have to do the bookkeeping to track `preempted` at all.